onset_maproom/agronomy.py

```diff
@@ -29,19 +31,84 @@ def soil_plant_water_bucket(
     peffective : DataArray
         effective precipitation today.
     et : DataArray
-        evapotransipiration of the plant today.
+        evapotranspiration of the plant today.
     taw : DataArray
         total available water that represents the maximum water capacity of the soil.
         
     Returns
     -------
     sm, drainage : Tuple of DataArray
         today soil moisture and drainage
+        
+    See Also
+    --------
+    soil_plant_water_balance
     
     """
     
     # Water Balance
-    wb = sm_yesterday + peffective - et
+    wb = (sm_yesterday + peffective - et).clip(min=0)
     drainage = (wb - taw).clip(min=0)
     sm = wb - drainage
+    return sm, drainage
+
+
+def soil_plant_water_balance(
+    peffective,
+    et,
+    taw,
+    sminit,
+    dim="T",
+):
+    """Compute soil-plant-water balance day after day over a growing season.
+    See `soil_plant_water_bucket` for the step by step algorithm definition.
+    
+    Parameters
+    ----------
+    peffective : DataArray
+        daily effective precipitation.
+    et : DataArray
+        daily evapotranspiration of the plant.
+    taw : DataArray
+        total available water that represents the maximum water capacity of the soil.
+    sminit : DataArray
+        timeless soil moisture to initialize the loop with.
+    dim : str, optional
+        daily time dimension to run the balance against (default `dim` ="T").
+        
+    Returns
+    -------
+    sm, drainage : Tuple of DataArray
+        daily soil moisture and drainage over the growing season.
+        
+    See Also
+    --------
+    soil_plant_water_bucket
+    
+    """
+    
+    # First Step
+    if np.size(et) == 1:
+        et = xr.DataArray(et)
+    sm0, drainage0 = soil_plant_water_bucket(
+        sminit,
+        peffective.isel({dim: 0}, drop=True),
+        et.isel({dim: 0}, missing_dims='ignore', drop=True),
+        taw,
+    )
+    # Give time dimension to sm and drainage    
+    sm = sm0.expand_dims(
+        {dim: peffective[dim].size}
+    ).assign_coords({dim: peffective[dim]}).copy()
```

Comment:
If I understand the docs correctly, you can collapse `expand_dims` and `assign_coords` into a single step:
```
sm = sm0.expand_dims({dim: peffective[dim]})
```

Why did you add the `copy`?

Reply:
if no copy, at some point, when trying to assign other time steps of sm, will get:

ValueError: Assignment destination is a view.  Do you want to .copy() array first?

And that remains true even after collapsing the two functions.